Components/Queues/Handler/PhpAmqpHandler.php

```diff
@@ -0,0 +1,232 @@
+            $channel->basic_ack($message->delivery_info['delivery_tag']);
+            --$this->expirationCount;
+        };
+
+        try {
+            $channel->basic_qos(null, 1, null);
+            $message = $channel->basic_consume($queueName, $consumerTag, false, false, false, false, $callback);
+            $this->isConsuming($channel);
+        } catch (\Exception $exception) {
+            $this->getExceptionHandler()($exception, ['headers' => $message->getHeaders(), 'body' => $message->getBody()]);
+            return;
+        }
+    }
+
+    /**
+     * Verifies if the channel is consuming a message.
+     * If there is a message to consume it calls the consume callback function
+     * If there is no message to consume it will put the worker in a wait state
+     * @param AMQPChannel $channel
+     * @throws \Exception
+     */
+    public function isConsuming(AMQPChannel $channel)
+    {
+        try {
+            while ($channel->is_consuming()) {
+                $channel->wait(null, true);
+            }
+        } catch (\Exception $exception) {
+            $this->log($exception->getMessage());
+            return;
+        }
+    }
+
+    /**
+     * Set the flag to stop the worker as true
+     * Dispatch the command to stop the work on next iteration
+     * @param string $consumerTag
+     * @throws \Exception
+     */
+    public function stopConsume(string $consumerTag)
+    {
+        try {
+            $this->stop();
+            if ($this->channel && $this->channel->getConnection()) {
+                $this->channel->basic_cancel($consumerTag, true, true);
+            }
+        } catch (\Exception $exception) {
+            $this->log($exception->getMessage());
+            exit(1);
+        }
+    }
+
+    /**
+     * Verifies if the message object is an instance of QueueMesseageInterface
+     * and if the handler is not an instance of QueueMessageHandlerInterface. Used by dispatchMessage function
+     * @param MessageInterface $message
+     * @return bool
+     */
+    private function isQueueMessage($message): bool
+    {
+        return $message instanceof QueueMessageInterface && !($this->endpoint->getHandler() instanceof QueueMessageHandlerInterface);
+    }
+
+    /**
+     * Log in console the message passed as param
+     * @param string $message
+     * @param array $ctx
+     * @throws \Exception
+     */
+    private function log(string $message, array $ctx = [])
+    {
+        if (null === $this->logger) {
+            return;
+        }
+
+        $now = new \DateTime();
+        $ctx['time'] = $now->format('Y-m-d H:i:s.u');
+        $this->logger->info($message, $ctx);
+    }
+
+    /**
+     * Dispatch the message to target system or events to database
+     * @param QueueMessage $message
+     * @throws \Exception
+     */
+    private function dispatchMessage(QueueMessage $message)
+    {
+        if (!$this->endpoint) {
+            $this->getExceptionHandler()('Endpoint is undefined');
```

Comment:
The exceptionHandler should only be used when you're dealing with incoming messages. Granted, the name is misleading, but it's main purpose is to requeue or rethrow the exception if the message fails to be deserialzed. 

If you don't declare it in your application, it rethrows the exception by default. In EAI, we override it with the `ReQueueSerializationExceptionHandler` which checks the type of exception and puts it in a different queue if the message is faulty.

In this case, and in most cases here, you just need to throw the exception. Only use this exception handler when you're deserializing the message.

Reply:
Ok, It's working with the exceptionHandler, but I'll refactory this soon.